lib/v3/parser.js

```diff
@@ -153,22 +154,17 @@ class Parser extends EventEmitter {
             item.defaultValue = variable.declarator.init.value;
         }
 
-        if (this.includeSourceLocations && variable.location) {
-            item.locations = [{
-                start: variable.location.start + parseContext.offset,
-                end: variable.location.end + parseContext.offset,
-            }];
-        }
+        this.attachLocationsIfRequired(item, variable, parseContext);
 
-        this.updateType(item);
+        updateType(item);
 
         this.emit('data', item);
     }
 
     emitMethodItem(method, parseContext, defaultVisibility, parentComment) {
         const comment = parentComment || utils.getCommentFromSourceCode(method.node, parseContext.sourceCode, { defaultVisibility });
```

Comment:
After refactoring this `parser` class is a manager that connect few parser and orchistrate they work. Mostly all communication now are abstract and we can provide different parsers. And that is very cool! However, `emit*` methods in that parser manager still working with raw `.node`. What do you think about that, how we can improve this moment and don't use raw ast nodes. In the future we reimplement `ScriptParser` and it will be very good if we can change only `ScriptParser` and keep common and template parser as it is.

Reply:
I like what you did with the sub-parser, especially the build method that instanciates the parser only once if necessary. I wanted to keep the state of the parser, but the way you did behaves more like a singleton and seems better.

For raw ast nodes, I suppose we could normalize them using a wrapper object so they would all be similar regardless of which parser they come from. There could be a common interface defined for wrapped nodes (it seems you already do that in some parts of the code where you create a new object from the original node). We just need to know what information to extract from the ast nodes.

Also, thanks for cleaning up the constants for events, I was going to do that before submitting but I forgot.

Reply:
> I like what you did with the sub-parser, especially the build method that instanciates the parser only once if necessary. I wanted to keep the state of the parser, but the way you did behaves more like a singleton and seems better.

Thanks. I think that is OK to make parsers as singleton there, due we create the new instances for each file that we parse, so that is a good to keep shared state, like top-level identifiers. I'm completelly not sure that is is a good idea to keep it into ScriptParser instead of GenericParser, but for now it's enough, so keep it as it is now.